heroku3/models/configvars.py

```diff
@@ -1,82 +1,108 @@
 
-    def __init__(self):
+    This class mutates the settings of the application as you mutate it's
+    content:
 
-        self.data = {}
-        self.app = None
-        self._h = None
+    .. code-block::
+
+       configuration = app.config()  # Assume app is an application
+       configure['SOME_SETTING'] = 'some value'
+
+    That last line of code will synchronously trigger a call to Heroku's
+    API to change the setting value. If this is not what you desire, you
+    can collect settings in a regular dictionary with::
+
+       import copy
+       configuration = app.config()  # Assume app is an application
+       safe_configuration = configuration.dict()
+       # mutate safe_configuration as you please
+       configuration.update(safe_configuration)
+
+    """
+
+    def __init__(self, data, app, h=None, **kwargs):
+        # assert app is not None
+        # copy: Ensures caller has no way to mutate our internal state outside
+        # our control
+        self.__data = copy.copy(data)
+        self.__app = app
+        self._h = h
 
         super(ConfigVars, self).__init__()
 
     def __repr__(self):
-        return repr(self.data)
+        return repr(self.__data)
 
     def __contains__(self, key):
-        v = None
-        try:
-            v = self[key]
-        except KeyError:
-            return False
-        else:
-            if v is None:
-                return False
-            else:
-                return True
+        return key in self.__data
 
     def __getitem__(self, key):
-        return self.data.get(key)
+        return self.__data.get(key)
 
     def __setitem__(self, key, value):
         # API expects JSON.
-        payload = json.dumps({key: value})
-
-        r = self._h._http_resource(
-            method='PATCH',
-            resource=('apps', self.app.name, 'config-vars'),
-            data=payload
-        )
-
-        self.data[key] = value
-        return r.ok
+        self.__data = self.__patch_config({key: value})
 
     def __delitem__(self, key):
-        data = self._h._resource_serialize({key: None})
-        r = self._h._http_resource(
-            method='PATCH',
-            resource=('apps', self.app.name, 'config-vars'),
-            data=data
-        )
+        data = self.__patch_config({key: None})
+        assert key not in data
+        self.__data = data
 
-        del self.data[key]
-        return r.ok
+    @property
+    def _resource(self):
+        return 'apps', self.__app.name, 'config-vars'
 
-    def update(self, newconf):
-        payload = self._h._resource_serialize(newconf)
+    def __patch_config(self, config):
+        payload = self._h._resource_serialize(config)
         r = self._h._http_resource(
             method='PATCH',
-            resource=('apps', self.app.name, 'config-vars'),
+            resource=self._resource,
             data=payload
         )
-
         r.raise_for_status()
```

Comment:
Just saw that `self._h._htts_resource()` actuall calls `r.raise_for_status()`. So that call is just not useful.